onnx/inliner/inliner.cc

```diff
@@ -0,0 +1,331 @@
+    }
+    // Otherwise, it is a reference to an outer-scope variable that should not be renamed.
+  }
+
+  template <bool isOutput>
+  void Bind(
+      google::protobuf::RepeatedPtrField<std::string>& formals,
+      const google::protobuf::RepeatedPtrField<std::string>& actuals) {
+    // Every formal parameter name FP should be replace by the corresponding actual parameter name AP.
+    // However, if AP is empty, it is a missing optional parameter. This does not make any difference
+    // for inputs. However, for outputs we use a unique dummy name to handle the case that it
+    // is used in an output-context where it is not optional.
+    ONNX_ASSERTM(
+        actuals.size() <= formals.size(), "Number of actual parameters cannot exceed number of formal parameters");
+    auto& current_scope = rename_scopes.back();
+    int i = 0;
+    for (; i < actuals.size(); ++i) {
+      std::string& formal = *formals.Mutable(i);
+      std::string rename_as = actuals.Get(i);
+      if (isOutput)
+        if (rename_as.empty())
+          rename_as = MakeUnique(formal);
+      current_scope[formal] = rename_as;
+      if (!rename_as.empty())
+        formal = rename_as;
+    }
+    for (; i < formals.size(); ++i) {
+      std::string& formal = *formals.Mutable(i);
+      std::string rename_as = isOutput ? MakeUnique(formal) : std::string("");
+      current_scope[formal] = rename_as;
+      if (!rename_as.empty())
+        formal = rename_as;
+    }
+  }
+
+  // Process a node:
+  void Transform(NodeProto& n) {
+    if (!n.name().empty())
+      n.set_name(MakeUnique(n.name()));
+
+    for (auto& x : *n.mutable_input()) {
+      LookupOrRename(x, false);
+    }
+    for (auto& y : *n.mutable_output()) {
+      LookupOrRename(y, true);
+    }
+    auto& attributes = *n.mutable_attribute();
+    for (auto attr_iter = attributes.begin(); attr_iter != attributes.end();) {
+      auto& attr = *attr_iter;
+      if (!attr.ref_attr_name().empty()) {
+        // Attribute-references must be replaced by the corresponding attribute-value in the call-node
+        // if the call-node contains the attribute. Otherwise, this attribute must be removed.
+        auto* attr_val = attr_map(attr.ref_attr_name());
+        if (attr_val != nullptr) {
+          // Copy value of attribute, but retain original name:
+          std::string name = attr.name();
+          attr = *attr_val;
+          attr.set_name(name);
+        } else {
+          attr_iter = attributes.erase(attr_iter);
+          continue;
+        }
+      }
+      // Subgraphs must be recursively processed.
+      if (attr.has_g()) {
+        Transform(*attr.mutable_g());
+      }
+      for (auto& graph : *attr.mutable_graphs())
+        Transform(graph);
+      ++attr_iter;
+    }
+  }
+
+  // Process a sub-graph, contained as an attribute in a control-flow op node.
+  void Transform(GraphProto& graph) {
+    rename_scopes.emplace_back();
+    for (auto& x : *graph.mutable_input())
```

Comment:
I would add a comment on how the new name is built to help developper debug.

Reply:
Done. See comments added to function `MakeUnique`.